pkg/controller/kubefedcluster/controller_test.go

```diff
@@ -38,71 +121,124 @@ func TestThresholdCheckedClusterStatus(t *testing.T) {
 	t4 := metav1.Time{Time: epoch.Add(4 * time.Second)}
 	t5 := metav1.Time{Time: epoch.Add(5 * time.Second)}
 
-	config := &util.ClusterHealthCheckConfig{
-		Period:           10 * time.Second,
-		FailureThreshold: 3,
-		SuccessThreshold: 1,
-		Timeout:          3 * time.Second,
-	}
+	It("validate controller informer actions and cluster data functions", func() {
+		ctx := context.TODO()
+		kubefedClusterSecret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: util.DefaultKubeFedSystemNamespace,
+				Name:      "test-cluster-data",
+			},
+			Data: map[string][]byte{
+				util.TokenKey: []byte("xxxxx"),
+			},
+		}
 
-	testCases := map[string]struct {
-		clusterStatus         *fedv1b1.KubeFedClusterStatus
-		storedClusterData     *ClusterData
-		expectedClusterStatus *fedv1b1.KubeFedClusterStatus
-		expectedResultRun     int64
-	}{
-		"ClusterReadyAtBegining": {
-			clusterStatus:         clusterStatus(corev1.ConditionTrue, t1, t1),
-			storedClusterData:     &ClusterData{clusterStatus: nil, resultRun: 0},
-			expectedClusterStatus: clusterStatus(corev1.ConditionTrue, t1, t1),
-			expectedResultRun:     1,
-		},
-		"ClusterNotReadyAtBegining": {
-			clusterStatus:         clusterStatus(corev1.ConditionFalse, t1, t1),
-			storedClusterData:     &ClusterData{clusterStatus: nil, resultRun: 0},
-			expectedClusterStatus: clusterStatus(corev1.ConditionFalse, t1, t1),
-			expectedResultRun:     1,
+		_, err := clientset.CoreV1().Secrets(util.DefaultKubeFedSystemNamespace).Create(kubefedClusterSecret)
+		Expect(err).ToNot(HaveOccurred())
+
+		kc := &fedv1b1.KubeFedCluster{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "cluster1",
+				Namespace: util.DefaultKubeFedSystemNamespace,
+				Labels: map[string]string{
+					"foo": "bar",
+				},
+			},
+			Spec: fedv1b1.KubeFedClusterSpec{
+				APIEndpoint: "https://my.example.com:80/path/to/endpoint",
+				SecretRef: fedv1b1.LocalSecretReference{
+					Name: kubefedClusterSecret.Name,
+				},
+			},
+			Status: *clusterStatus(corev1.ConditionTrue, metav1.Time{Time: metav1.Now().Add(1 * time.Second)}, metav1.Time{Time: metav1.Now().Add(1 * time.Second)}),
+		}
+
+		err = k8sClient.Create(ctx, kc)
+		Expect(err).ToNot(HaveOccurred())
+
+		cc.addToClusterSet(kc)
+		Expect(err).ToNot(HaveOccurred())
+		_, found := cc.clusterDataMap[kc.Name]
+		Expect(found).To(BeTrue())
+
+		cc.delFromClusterSet(kc)
+		_, found = cc.clusterDataMap[kc.Name]
+		Expect(found).NotTo(BeTrue())
+
+		fedCluster := &fedv1b1.KubeFedCluster{}
+		err = k8sClient.Get(ctx, client.ObjectKey{Namespace: util.DefaultKubeFedSystemNamespace, Name: kc.Name}, fedCluster)
+		Expect(err).ToNot(HaveOccurred())
+
+		fedCluster.Spec.APIEndpoint = "https://my.example.com:80/path/to/newendpoint"
+		err = k8sClient.Update(ctx, fedCluster)
+		Expect(err).ToNot(HaveOccurred())
+
+		Eventually(func() bool {
+			_, found = cc.clusterDataMap[kc.Name]
+			return found
+		}, 30*time.Second).Should(BeTrue())
```

Comment:
Recommend: Verify that the data updated (API Endpoint) equals what is updated above?

Reply:
👍